ci-operator/step-registry/gcp/provision/static-creds/gcp-provision-static-creds-commands.sh

```diff
@@ -0,0 +1,213 @@
+then
+  gcloud auth activate-service-account --key-file="${GCP_SHARED_CREDENTIALS_FILE}"
+  gcloud config set project "${GOOGLE_PROJECT_ID}"
+fi
+
+# For OCP 4.15+
+CCO_CUSTOM_ROLE="projects/${GOOGLE_PROJECT_ID}/roles/installer_qe_cco_permissions"
+IMAGE_REGISTRY_CUSTOM_ROLE="projects/${GOOGLE_PROJECT_ID}/roles/installer_qe_image_registry_permissions"
+
+CLUSTER_NAME="${NAMESPACE}-${UNIQUE_HASH}"
+MPREFIX="${SHARED_DIR}/manifest"
+creds_requests_dir="$(mktemp -d)"
+creds_keys_dir="$(mktemp -d)"
+creds_manifests_dir="$(mktemp -d)"
+
+echo "$(date -u --rfc-3339=seconds) - Extracting GCP credentials requests from the release image..."
+ADDITIONAL_OC_EXTRACT_ARGS=""
+if [[ "${EXTRACT_MANIFEST_INCLUDED}" == "true" ]]; then
+  ADDITIONAL_OC_EXTRACT_ARGS="${ADDITIONAL_OC_EXTRACT_ARGS} --included --install-config=${SHARED_DIR}/install-config.yaml"
+fi
+echo "OC Version:"
+which oc
+oc version --client
+oc adm release extract --help
+
+dir=$(mktemp -d)
+pushd "${dir}"
+cp ${CLUSTER_PROFILE_DIR}/pull-secret pull-secret
+oc registry login --to pull-secret
+oc adm release extract --registry-config pull-secret --credentials-requests --cloud=gcp --to="${creds_requests_dir}" ${ADDITIONAL_OC_EXTRACT_ARGS} "${TESTING_RELEASE_IMAGE}"
+rm pull-secret
+popd
+
+if [ X"${FEATURE_SET}" != X"TechPreviewNoUpgrade" ]; then
+  # workaround
+  # Bug 2035903 - One redundant capi-operator credential requests in “oc adm extract --credentials-requests”
+  # https://bugzilla.redhat.com/show_bug.cgi?id=2035903
+  echo "$(date -u --rfc-3339=seconds) - WARN: Workaround for https://bugzilla.redhat.com/show_bug.cgi?id=2035903, removing TechPreviewNoUpgrade CredentialsRequests"
+  annotation="TechPreviewNoUpgrade"
+  remove_tp_creds_requests "${creds_requests_dir}" "${annotation}" || exit 1
+fi
+
+echo "$(date -u --rfc-3339=seconds) - Extracted GCP credentials requests to directory: ${creds_requests_dir}"
+ls "${creds_requests_dir}"
+
+echo "$(date -u --rfc-3339=seconds) - Creating GCP IAM service accounts for CCO manual mode..."
+sa_suffix=${GCP_SERVICE_ACCOUNT#*@}
+for yaml_filename in $(ls -p "${creds_requests_dir}"/*.yaml | awk -F'/' '{print $NF}'); do
+  echo "$(date -u --rfc-3339=seconds) - Processing ${yaml_filename}"
+  readarray -t roles < <(yq-go r "${creds_requests_dir}/${yaml_filename}" "spec.providerSpec.predefinedRoles" | sed 's/- //g')
+  secret_name=$(yq-go r "${creds_requests_dir}/${yaml_filename}" "spec.secretRef.name")
+  secret_namespace=$(yq-go r "${creds_requests_dir}/${yaml_filename}" "spec.secretRef.namespace")
+  metadata_name=$(yq-go r "${creds_requests_dir}/${yaml_filename}" "metadata.name")
+  if [ -z "$secret_name" ] || [ -z "$secret_namespace" ] || [ -z "$metadata_name" ]; then
+    echo "$(date -u --rfc-3339=seconds) - ERROR: spec.secretRef.name, spec.secretRef.namespace or metadata.name is empty"
+    cat "${creds_requests_dir}/${yaml_filename}"
+    exit 1
+  fi
+  if [ -z "${roles:-}" ]; then
```

Comment:
Why we need such special operator for `cloud-credential-operator` and `openshift-image-registry`, is there offical doc to mention this?

Reply:
I didn't search for official doc yet, but noticed that "spec.providerSpec.**predefinedRoles**" is absent in the 2 credentials requests YAML files, for OCP 4.15 & 4.16. So I pre-configured custom roles according to the "spec.providerSpec.**permissions**" section in these 2 YAML files.

Reply:
Can I understand the **predefinedRoles** is a pre-configured set of **permissions**? Can we directly grant the new users with the listed "spec.providerSpec.**permissions**" ?

Reply:
It looks no way to grant permission(s) directly to IAM service accounts. Instead, we can configure custom role(s) to include the interested permission(s) and then grant the custom role to IAM service accounts.

Reply:
If that, I prefer to create the custom role on demand with the requested permissions, once that, we can ensure we always using the correct permission to run our testing version by version (to avoid some change when some change happened on different versions)

Reply:
@jianlinliu The permanent deletion of a custom role takes 30 days (see [Google doc](https://cloud.google.com/iam/docs/creating-custom-roles?_ga=2.17178032.-881852630.1701826900)), and it seems the upper limit is 300 custom roles per project. So I'm afraid dynamically creating & deleting custom roles would trigger the limit. Your suggestion? Thanks!

Reply:
Good point.